<a href="https://colab.research.google.com/github/slouie88/pandas-practice/blob/master/Python_pandas_Practice_Problems_UC_Berkeley_MIDS_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python pandas Practice Problems

Welcome to datascience@berkeley's collection of Python pandas practice problems.

To use this notebook, go to `File` --> `Save a copy in Drive` in the menu bar above to create a copy of this notebook on your computer. A new tab will open with the copy and you will be the owner. Your version will not be visible to anyone else.

A few notes to get you started:

* The space provided for code is not a line limit or hint. Add and use as many lines of code or additional cells as necessary.
* Lines preceded with '#' are comments and provide additional explanation but will not affect written code or execution.
* Solutions can be found in the complementary [Python pandas Practice Problems for Beginners - Solutions notebook](https://colab.research.google.com/drive/1f8PTrhgJR093fBth5Ka23KCWTIGbgPds?usp=sharing).
* Additional resources for the concepts used in each problem can be found at [the blog post here](https://ischoolonline.berkeley.edu/blog/python-pandas-practice-problems/).


In [1]:
# Import statements go here
import math
import pandas as pd
import statsmodels.api as sm

## Importing Data and Making a DataFrame
The statsmodels package (installed in the code cell above) includes built-in datasets. Execute the code below to download data from the [American National Election Studies of 1996](https://www.statsmodels.org/dev/datasets/generated/anes96.html) and print a detailed description of the schema.

The next cell extracts the `Dataset` object from the submodule and saves the `DataFrame` to the variable `df`. In the questions that follow, use `df` when referencing the dataset.

In [2]:
anes96 = sm.datasets.anes96
print(anes96.NOTE)

::

    Number of observations - 944
    Number of variables - 10

    Variables name definitions::

            popul - Census place population in 1000s
            TVnews - Number of times per week that respondent watches TV news.
            PID - Party identification of respondent.
                0 - Strong Democrat
                1 - Weak Democrat
                2 - Independent-Democrat
                3 - Independent-Indpendent
                4 - Independent-Republican
                5 - Weak Republican
                6 - Strong Republican
            age : Age of respondent.
            educ - Education level of respondent
                1 - 1-8 grades
                2 - Some high school
                3 - High school graduate
                4 - Some college
                5 - College degree
                6 - Master's degree
                7 - PhD
            income - Income of household
                1  - None or less than $2,999
                2  - $3,000-$4,9

In [3]:
dataset_anes96 = anes96.load_pandas()
df = dataset_anes96.data

## 1. DataFrame Basic Properties Exercise

Our DataFrame (`df`) contains data on registered voters in the United States, including demographic information and political preference. Using `pandas`, print the first 5 rows of the DataFrame to get a sense of what the data looks like. Next, answer the following questions:


*   How many observations are in the DataFrame?
    - 944 observations
*   How many variables are measured (how many columns)?
    - 11 variables
*   What is the age of the youngest person in the data? The oldest?
    - Youngest age: 19yrs old
    - Oldest age: 91 yrs old
*   How many days a week does the average respondent watch TV news (round to the nearest tenth)?
    - 3.7 times
*   Check for missing values. Are there any?
    - None






In [4]:
# Your code here
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  educ  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0   3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0   4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0   6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0   6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0   6.0     1.0   0.0   

   logpopul  
0 -2.302585  
1  5.247550  
2  3.437208  
3  4.420045  
4  6.461624

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   popul     944 non-null    float64
 1   TVnews    944 non-null    float64
 2   selfLR    944 non-null    float64
 3   ClinLR    944 non-null    float64
 4   DoleLR    944 non-null    float64
 5   PID       944 non-null    float64
 6   age       944 non-null    float64
 7   educ      944 non-null    float64
 8   income    944 non-null    float64
 9   vote      944 non-null    float64
 10  logpopul  944 non-null    float64
dtypes: float64(11)
memory usage: 81.2 KB


In [6]:
youngest = df["age"].min()
oldest = df["age"].max()
print(f"Age of the youngest person observed in the dataset: {youngest}")
print(f"Age of the oldest person observed in the dataset: {oldest}")

Age of the youngest person observed in the dataset: 19.0
Age of the oldest person observed in the dataset: 91.0


In [7]:
average_tv_per_week = df["TVnews"].mean()
print(f"The average person watches TV approximately {round(average_tv_per_week, 1)} times per week")

The average person watches TV approximately 3.7 times per week


## 2. Data Processing Exercise

We want to adjust the dataset for our use. Do the following:


*   Rename the `educ` column `education`.
*   Create a new column called `party` based on each respondent's answer to `PID`. `party` should equal `Democrat` if the respondent selected either Strong Democrat or Weak Democrat. `party` will equal `Republican` if the respondent selected Strong or Weak Republican for `PID` and `Independent` if they selected anything else.
*   Create a new column called `age_group` that buckets respondents into the following categories based on their `age`: 18-24, 25-34, 35-44, 45-54, 55-64, and 65 and over.



In [8]:
# Your code here
df.rename(columns={"educ": "education"}, inplace=True)
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul  
0 -2.302585  
1  5.247550  
2  3.437208  
3  4.420045  
4  6.461624

In [9]:
party = []
for i in df["PID"]:
    if i == 0 or i == 1:
        party.append("Democrat")
    elif i == 5 or i == 6:
        party.append("Republican")
    else:
        party.append("Independent")
df["party"] = party
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul       party  
0 -2.302585  Republican  
1  5.247550    Democrat  
2  3.437208    Democrat  
3  4.420045    Democrat  
4  6.461624    Democrat

In [10]:
age_group = []
for age in df["age"]:
    if age >= 18 and age < 25:
        age_group.append("18-24")
    elif age >= 25 and age < 35:
        age_group.append("25-35")
    elif age >= 35 and age < 45:
        age_group.append("35-45")
    elif age >= 45 and age < 55:
        age_group.append("45-55")
    elif age >= 55 and age < 65:
        age_group.append("55-65")
    else:
        age_group.append("65 and over")
df["age_group"] = age_group
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul       party    age_group  
0 -2.302585  Republican        35-45  
1  5.247550    Democrat        18-24  
2  3.437208    Democrat        18-24  
3  4.420045    Democrat        25-35  
4  6.461624    Democrat  65 and over

## 3. Filtering Data Exercise

Use the filtering method to find all the respondents who have the impression that Bill Clinton is moderate or conservative (`ClinLR` equals 4 or higher). How many respondents are in this subset?

Among these respondents, how many have a household income less than $50,000 and attended at least some college?

In [11]:
# Your code here
sub_df = df[df["ClinLR"] >= 4]
sub_df

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  \
3     83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   
4    640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   
5    110.0     3.0     3.0     4.0     6.0  1.0  21.0        4.0     1.0   
6    100.0     7.0     5.0     6.0     4.0  1.0  77.0        4.0     1.0   
7     31.0     1.0     5.0     4.0     5.0  4.0  21.0        4.0     1.0   
..     ...     ...     ...     ...     ...  ...   ...        ...     ...   
911   24.0     7.0     3.0     4.0     5.0  1.0  38.0        7.0    24.0   
912   18.0     0.0     2.0     4.0     6.0  1.0  51.0        7.0    24.0   
921    0.0     7.0     2.0     4.0     5.0  2.0  53.0        6.0    24.0   
925    0.0     0.0     3.0     4.0     4.0  0.0  48.0        6.0    24.0   
933   27.0     7.0     4.0     4.0     7.0  2.0  46.0        3.0    24.0   

     vote  logpopul        party    age_group  
3     0.0  4.420045     Democrat        25-35  
4     0.0  6.461624     Democrat  65 and over  
5     0.0  4.701389     Democrat        18-24  
6     0.0  4.606170     Democrat  65 and over  
7     0.0  3.437208  Independent        18-24  
..    ...       ...          ...          ...  
911   0.0  3.182212     Democrat        35-45  
912   0.0  2.895912     Democrat        45-55  
921   0.0 -2.302585  Independent        45-55  
925   0.0 -2.302585     Democrat        45-55  
933   0.0  3.299534  Independent        45-55  

[282 rows x 13 columns]

In [12]:
sub_df2 = sub_df[(sub_df["income"] < 20) & (sub_df["education"] >= 4)]
sub_df2

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  \
3     83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   
4    640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   
5    110.0     3.0     3.0     4.0     6.0  1.0  21.0        4.0     1.0   
6    100.0     7.0     5.0     6.0     4.0  1.0  77.0        4.0     1.0   
7     31.0     1.0     5.0     4.0     5.0  4.0  21.0        4.0     1.0   
..     ...     ...     ...     ...     ...  ...   ...        ...     ...   
553   35.0     0.0     3.0     4.0     6.0  1.0  36.0        7.0    19.0   
563    0.0     1.0     4.0     4.0     6.0  1.0  41.0        4.0    19.0   
568   71.0     2.0     4.0     6.0     2.0  0.0  57.0        4.0    19.0   
570   83.0     4.0     1.0     4.0     6.0  2.0  43.0        6.0    19.0   
571  360.0     3.0     4.0     4.0     7.0  0.0  34.0        4.0    19.0   

     vote  logpopul        party    age_group  
3     0.0  4.420045     Democrat        25-35  
4     0.0  6.461624     Democrat  65 and over  
5     0.0  4.701389     Democrat        18-24  
6     0.0  4.606170     Democrat  65 and over  
7     0.0  3.437208  Independent        18-24  
..    ...       ...          ...          ...  
553   0.0  3.558201     Democrat        35-45  
563   0.0 -2.302585     Democrat        35-45  
568   0.0  4.264087     Democrat        55-65  
570   0.0  4.420045  Independent        35-45  
571   0.0  5.886382     Democrat        25-35  

[98 rows x 13 columns]

In [13]:
print(f"There are {len(sub_df)} respondents who have the impression that Bill Clinton is moderate or conservative")
print(f"Out of the {len(sub_df)} respondents, {len(sub_df2)} of them have a household income of less than $50,000 and have attended at least some college")

There are 282 respondents who have the impression that Bill Clinton is moderate or conservative
Out of the 282 respondents, 98 of them have a household income of less than $50,000 and have attended at least some college


## 4. Calculating From Data Exercise

For each of the below match-ups, choose the group that is more likely to vote for Bill Clinton. You can calculate this using the percentage of each group that intends to vote for Clinton (`vote`). Which match-up was the closest? Which had the biggest difference?

Another way to think about this: Given that a respondent is a Democrat, there is a ____ percent chance they will vote for Clinton. How does this value change if the respondent is a Republican?

*   Democrats or Republicans
    - Democrats
*   People younger than 44 or People 44 and older
    - People younger than 44
*   People who watch TV news at least 6 days a week or People who watch TV news less than 3 days a week
    - People who watch TV news at least 6 days a week
*   People who live somewhere with a population greater than the average respondent or People who live in a place with a population equal to or less than the average respondent
    - People who live with somewhere with a population greater than the average rrespondent


In [14]:
# Your code here
democrats = df[df["party"] == "Democrat"]
vote_d_c = len([v for v in democrats["vote"] if v == 0])
total_votes_d_c = len(democrats["vote"])
vote_d_c_pct = vote_d_c / total_votes_d_c
print(f"The percentage of Democrats voting for Clinton is: {round(vote_d_c_pct * 100, 2)}%")

republicans = df[df["party"] == "Republican"]
vote_r_c = len([v for v in republicans["vote"] if v == 0])
total_votes_r_c = len(republicans["vote"])
vote_r_c_pct = vote_r_c / total_votes_r_c
print(f"The percentage of Republicans voting for Clinton is: {round(vote_r_c_pct * 100, 2)}%")

The percentage of Democrats voting for Clinton is: 96.32%
The percentage of Republicans voting for Clinton is: 10.46%


In [15]:
younger_than_44 = df[df["age"] < 44]
vote_44less_c = len([v for v in younger_than_44["vote"] if v == 0])
total_votes_44less_c = len(younger_than_44["vote"])
vote_44less_c_pct = vote_44less_c / total_votes_44less_c
print(f"The percentage of respondents younger than 44 voting for Clinton is: {round(vote_44less_c_pct * 100, 2)}%")

older_than_44 = df[df["age"] >= 44]
vote_44more_c = len([v for v in older_than_44["vote"] if v == 0])
total_votes_44more_c = len(older_than_44["vote"])
vote_44more_c_pct = vote_44more_c / total_votes_44more_c
print(f"The percentage of respondents older than or aged 44 voting for Clinton is: {round(vote_44more_c_pct * 100, 2)}%")

The percentage of respondents younger than 44 voting for Clinton is: 59.48%
The percentage of respondents older than or aged 44 voting for Clinton is: 57.29%


In [16]:
tv_6more = df[df["TVnews"] >= 6]
vote_6more_c = len([v for v in tv_6more["vote"] if v == 0])
total_votes_6more_c = len(tv_6more["vote"])
vote_6more_c_pct = vote_6more_c / total_votes_6more_c
print(f"The percentage of respondents who watch tv news at least 6 days per week voting for Clinton is: {round(vote_6more_c_pct * 100, 2)}%")

tv_3less = df[df["TVnews"] < 3]
vote_3less_c = len([v for v in tv_3less["vote"] if v == 0])
total_votes_3less_c = len(tv_3less["vote"])
vote_3less_c_pct = vote_3less_c / total_votes_3less_c
print(f"The percentage of respondents who watch tv news at least 6 days per week voting for Clinton is: {round(vote_3less_c_pct * 100, 2)}%")

The percentage of respondents who watch tv news at least 6 days per week voting for Clinton is: 57.81%
The percentage of respondents who watch tv news at least 6 days per week voting for Clinton is: 55.5%


In [17]:
pop_mean_greater = df[df["popul"] > df["popul"].mean()]
vote_mean_greater_c = len([v for v in pop_mean_greater["vote"] if v == 0])
total_votes_mean_greater_c = len(pop_mean_greater["vote"])
vote_mean_greater_c_pct = vote_mean_greater_c / total_votes_mean_greater_c
print(f"The percentage of respondents who live in an above average population area voting for Clinton is: {round(vote_mean_greater_c_pct * 100, 2)}%")

pop_mean_less = df[df["popul"] <= df["popul"].mean()]
vote_mean_less_c = len([v for v in pop_mean_less["vote"] if v == 0])
total_votes_mean_less_c = len(pop_mean_less["vote"])
vote_mean_less_c_pct = vote_mean_less_c / total_votes_mean_less_c
print(f"The percentage of respondents who live in an above average population area voting for Clinton is: {round(vote_mean_less_c_pct * 100, 2)}%")

The percentage of respondents who live in an above average population area voting for Clinton is: 72.34%
The percentage of respondents who live in an above average population area voting for Clinton is: 55.92%


## 5. Grouping Data Exercise

Use the `groupby()` method to bucket respondents by `age_group`. Which age group is the most conservative? Which watches TV news the least?

Next, calculate 5 percentile groups based on income. Group the dataset by these percentiles. Which income bracket is the most liberal? Which is the most conservative? The oldest? Highest educated?

In [18]:
# Your code here
grouped_age_group = df.groupby("age_group")
grouped_age_group[["selfLR", "TVnews"]].mean()

selfLR    TVnews
age_group                      
18-24        4.000000  2.358491
25-35        4.217391  2.570652
35-45        4.257143  3.118367
45-55        4.285714  3.916667
55-65        4.532258  4.516129
65 and over  4.529412  5.523529

In [19]:
quantile_ranges = [0.2, 0.4, 0.6, 0.8, 1.0]
income_quantiles = df["income"].quantile(q=quantile_ranges)
income_quantile = []
for i in df["income"]:
    if i <= income_quantiles[0.2]:
        income_quantile.append("Q1")
    elif i <= income_quantiles[0.4]:
        income_quantile.append("Q2")
    elif i <= income_quantiles[0.6]:
        income_quantile.append("Q3")
    elif i <= income_quantiles[0.8]:
        income_quantile.append("Q4")
    else:
        income_quantile.append("Q5")
df["income_quantile"] = income_quantile

grouped_income_quantile = df.groupby("income_quantile")
grouped_income_quantile[["selfLR", "age", "education"]].mean()

selfLR        age  education
income_quantile                                
Q1               4.282297  50.009569   3.722488
Q2               4.295567  48.266010   4.246305
Q3               4.099379  45.583851   4.571429
Q4               4.497537  45.128079   4.862069
Q5               4.422619  45.589286   5.636905

## 6. Voting Across the Aisle

We are interested in learning more about respondents who's political views differ strongly from the candidate they expect to vote for. Using `selfLR`, `vote`, `ClinLR`, and `DoleLR`, work through the following questions. Your interpretation may differ from the answer key.

*   What is the largest recorded difference between a respondent's political leaning and their impression of their intended candidate's political leaning?
*   How many respondents exhibit a difference of that magnitude?
*   Make a separate DataFrame called `sway` that only includes voters who exhibit a difference greater than |3|.
*   Among those in `sway`, are respondents more likely to be voting for a candidate more conservative or more liberal than their own political leaning?
*   In `sway`, which candidate is the more popular choice?



In [20]:
# Your code here
diffLR = []
for i, r in df.iterrows():
    if r["vote"] == 0:
        diff = r["selfLR"] - r["ClinLR"]
    else:
        diff = r["selfLR"]- r["DoleLR"]
    diffLR.append(diff)
df["diffLR"] = diffLR

max_diff = df["diffLR"].abs().max()
print(f"The largest recorded difference between a respondent's political leaning and their impression of their intended candidate's political leaning is {max_diff}")

The largest recorded difference between a respondent's political leaning and their impression of their intended candidate's political leaning is 6.0


In [21]:
grouped_diffLR = df.groupby("diffLR")
grouped_diffLR["diffLR"].count()

diffLR
-4.0      2
-3.0     14
-2.0     84
-1.0    216
 0.0    319
 1.0    200
 2.0     69
 3.0     35
 4.0      3
 5.0      1
 6.0      1
Name: diffLR, dtype: int64

In [22]:
sway = df[df["diffLR"].abs() >= 3]
sway.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
29  450.0     3.0     4.0     7.0     1.0  0.0  34.0        3.0     2.0   0.0   
46    0.0     7.0     4.0     3.0     1.0  6.0  45.0        3.0     3.0   1.0   
77   32.0     5.0     3.0     7.0     4.0  1.0  65.0        1.0     5.0   0.0   
78  390.0     7.0     3.0     6.0     2.0  2.0  24.0        3.0     5.0   0.0   
88   31.0     2.0     3.0     4.0     6.0  6.0  20.0        4.0     6.0   1.0   

    logpopul        party    age_group income_quantile  diffLR  
29  6.109470     Democrat        25-35              Q1    -3.0  
46 -2.302585   Republican        45-55              Q1     3.0  
77  3.468856     Democrat  65 and over              Q1    -4.0  
78  5.966403  Independent        18-24              Q1    -3.0  
88  3.437208   Republican        18-24              Q1    -3.0

In [23]:
total_sway = sway["diffLR"].count()

sway_conservative = sway[sway["diffLR"] < 0]
pct_conservative = sway_conservative["diffLR"].count() / total_sway
print(f"Among those in sway, the percentage of respondents voting more conservative than their political leaning is {round(pct_conservative * 100, 2)}%")

sway_liberal = sway[sway["diffLR"] >= 0]
pct_liberal = sway_liberal["diffLR"].count() / total_sway
print(f"Among those in sway, the percentage of respondents voting more liberal than their political leaning is {round(pct_liberal * 100, 2)}%")

Among those in sway, the percentage of respondents voting more conservative than their political leaning is 28.57%
Among those in sway, the percentage of respondents voting more liberal than their political leaning is 71.43%


In [24]:
sway_clin = sway[sway["vote"] == 0]
pct_clin = sway_clin["vote"].count() / total_sway
print(f"Among those in sway, the percentage of respondents voting more Clinton is {round(pct_clin * 100, 2)}%")

sway_dole = sway[sway["vote"] == 1]
pct_dole = sway_dole["vote"].count() / total_sway
print(f"Among those in sway, the percentage of respondents voting more Dole is {round(pct_dole * 100, 2)}%")


Among those in sway, the percentage of respondents voting more Clinton is 85.71%
Among those in sway, the percentage of respondents voting more Dole is 14.29%


# BSD 3-Clause License

*Copyright (c) 2022, UC Berkeley School of Information*

*All rights reserved.*

*Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:*

*1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.*

*2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.*

*3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.*

*THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.*